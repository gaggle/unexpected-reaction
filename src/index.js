const { mount, Simulate } = require("react-dom-testing");

module.exports = {
  name: "unexpected-reaction",
  installInto: expect => {
    expect.addType({
      name: "ReactElement",
      base: "wrapperObject",
      unwrap: value => mount(value),
      prefix: output => output,
      suffix: output => output,
      identify: value =>
        typeof value === "object" &&
        value !== null &&
        (typeof value.type === "function" || typeof value.type === "string") &&
        typeof value.hasOwnProperty === "function" &&
        value.hasOwnProperty("props") &&
        value.hasOwnProperty("ref") &&
        value.hasOwnProperty("key")
    });

    expect.addAssertion(
      "<ReactElement> when mounted <assertion>",
      (expect, subject) => {
        expect.errorMode = "bubble";
        return expect.shift(mount(subject));
      }
    );

    expect.addAssertion(
      "<DOMElement> to [exhaustively] satisfy <ReactElement>",
      (expect, subject, value) => {
        expect.errorMode = "bubble";
        return expect(subject, "to [exhaustively] satisfy", mount(value));
      }
    );

    expect.addAssertion(
      "<DOMElement> with (event|events) <array|object|string> <assertion?>",
      (expect, subject, value) => {
        expect.errorMode = "nested";

        []
          .concat(value)
          .map(event => (typeof event === "string" ? { type: event } : event))
          .forEach(event => {
            let target = subject;

            if (event.target) {
              expect(subject, "to contain elements matching", event.target);
              target = subject.querySelector(event.target);
            }

            if (event.type === "change" && typeof event.value === "string") {
              target.value = event.value;
            }

            if (!Simulate[event.type]) {
              expect.fail(
                `Event '${
                  event.type
                }' is not supported by Simulate\nSee https://reactjs.org/docs/events.html#supported-events`
              );
            }

            Simulate[event.type](target, event.data);
          });

        return expect.shift(subject);
      }
    );
  }
};
