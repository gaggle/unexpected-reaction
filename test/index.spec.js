import React, { Component } from "react";
import PropTypes from "prop-types";

import expect from "./unexpected-with-plugins";
import { mount, simulate } from "../src/";

class Hello extends Component {
  render() {
    const { children, ...other } = this.props;

    return (
      <div {...other}>
        <div className="label">Hello:</div>
        <div className="value" data-test="value">
          {children}
        </div>
      </div>
    );
  }
}

Hello.propTypes = {
  children: PropTypes.node.isRequired
};

const Stateless = ({ className, children }) => (
  <Hello className={className}>{children}</Hello>
);

Stateless.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

class Toggle extends Component {
  constructor(props) {
    super(props);

    this.state = {
      checked: false
    };

    this.onClick = () => {
      this.setState(state => ({
        checked: !state.checked
      }));
    };
  }

  render() {
    const { checked } = this.state;

    return (
      <div onClick={this.onClick}>
        {checked && <span style={{ color: "green" }}>TRUE</span>}
        {!checked && <span style={{ color: "red" }}>FALSE</span>}
      </div>
    );
  }
}

class MyInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: "",
      selected: ""
    };

    this.onKeyDown = data => {
      if (data.keyCode === 13) {
        // Enter
        this.setState(state => ({
          value: "",
          selected: state.value
        }));
      }
    };

    this.onChange = data => {
      this.setState({
        value: data.target.value
      });
    };
  }

  render() {
    return (
      <fieldset>
        <label htmlFor="name">Enter your name:</label>
        <input
          type="input"
          name="name"
          value={this.state.value}
          onChange={this.onChange}
          onKeyDown={this.onKeyDown}
        />
        <span data-test="name">{this.state.selected}</span>
      </fieldset>
    );
  }
}

describe("unexpected-reaction", () => {
  describe("ReactElement", () => {
    describe("when mounted", () => {
      it("renders the given component into the DOM and forwards the node to the next assertion", () => {
        expect(
          <Hello>Jane Doe</Hello>,
          "when mounted",
          "queried for first",
          "[data-test=value]",
          "to have text",
          "Jane Doe"
        );
      });

      describe("on a stateless component", () => {
        it("renders the given component into the DOM and forwards the node to the next assertion", () => {
          expect(
            <Stateless className="fancy">I am stateless</Stateless>,
            "when mounted",
            "queried for first",
            "[data-test=value]",
            "to have text",
            "I am stateless"
          );
        });
      });
    });
  });

  describe("DOMElement", () => {
    describe("to satisfy", () => {
      it("mounts the given ReactElement and satisfies the subject against it", () => {
        expect(
          <Hello>Jane Doe</Hello>,
          "when mounted",
          "to satisfy",
          <div>
            <div className="label">Hello:</div>
            <div>Jane Doe</div>
          </div>
        );
      });

      it("fails with a diff", () => {
        expect(() => {
          expect(
            <Hello>Jane Doe</Hello>,
            "when mounted",
            "to satisfy",
            <div>
              <div className="label">Hello:</div>
              <div>Jane Doe!</div>
            </div>
          );
        }, "with error matching snapshot");
      });

      describe("on conditional tree", () => {
        it("mounts the given ReactElement and satisfies the subject against it", () => {
          const mounted = mount(<Toggle />);

          expect(
            mounted,
            "to satisfy",
            <div>
              <span>FALSE</span>
            </div>
          );

          simulate(mounted, "click");

          expect(
            mounted,
            "to satisfy",
            <div>
              <span>TRUE</span>
            </div>
          );
        });
      });
    });

    describe("to exhaustivily satisfy", () => {
      it("mounts the given ReactElement and satisfies the subject against it", () => {
        expect(
          <Hello data-test-id="hello">Jane Doe</Hello>,
          "when mounted",
          "to exhaustively satisfy",
          <div data-test-id="hello">
            <div className="label">Hello:</div>
            <div className="value" data-test="value">
              Jane Doe
            </div>
          </div>
        );
      });

      it("fails with a diff", () => {
        expect(() => {
          expect(
            <Hello data-test-id="hello">Jane Doe</Hello>,
            "when mounted",
            "to exhaustively satisfy",
            <div>
              <div className="label">Hello:</div>
              <div className="value" data-test="value">
                Jane Doe
              </div>
            </div>
          );
        }, "with error matching snapshot");
      });
    });

    describe("with event", () => {
      describe("when given a string", () => {
        it("invokes the given event type on the subject", () => {
          expect(
            <Toggle />,
            "when mounted",
            "with event",
            "click",
            "to have text",
            "TRUE"
          );
        });
      });

      describe("when given an object", () => {
        expect(
          <MyInput />,
          "when mounted",
          "with event",
          {
            type: "change",
            value: "Jane Doe",
            target: "input"
          },
          "queried for first",
          "input",
          "to have attribute",
          {
            value: "Jane Doe"
          }
        );
      });

      describe("when given an array", () => {
        it("applies all of the events", () => {
          expect(
            <MyInput />,
            "when mounted",
            "with events",
            [
              {
                type: "change",
                value: "Jane Doe",
                target: "input"
              },
              {
                type: "keyDown",
                target: "input",
                data: {
                  keyCode: 13
                }
              }
            ],
            "queried for first",
            "[data-test=name]",
            "to have text",
            "Jane Doe"
          );
        });

        it("accepts string and object events", () => {
          expect(
            <Toggle />,
            "when mounted",
            "with events",
            ["click", { type: "click" }],
            "to have text",
            "FALSE"
          );
        });
      });

      it("fails if it cant find the event target", () => {
        expect(() => {
          expect(
            <Toggle />,
            "when mounted",
            "with event",
            { type: "click", target: ".foobar" },
            "to have text",
            "TRUE"
          );
        }, "with error matching snapshot");
      });

      it("fails if the event type is not known", () => {
        expect(() => {
          expect(
            <Toggle />,
            "when mounted",
            "with event",
            "press",
            "to have text",
            "TRUE"
          );
        }, "with error matching snapshot");
      });
    });
  });
});
