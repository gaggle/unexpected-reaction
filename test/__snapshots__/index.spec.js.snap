// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unexpected-reaction DOMElement to exhaustivily satisfy fails with a diff 1`] = `
"expected
<div data-test-id=\\"hello\\">
  <div class=\\"label\\">Hello:</div>
  <div class=\\"value\\" data-test=\\"value\\">Jane Doe</div>
</div>
to exhaustively satisfy
<div>
  <div class=\\"label\\">Hello:</div>
  <div class=\\"value\\" data-test=\\"value\\">Jane Doe</div>
</div>

<div
  data-test-id=\\"hello\\" // should be removed
><div class=\\"label\\">Hello:</div><div class=\\"value\\" data-test=\\"value\\">Jane Doe</div></div>"
`;

exports[`unexpected-reaction DOMElement to satisfy fails with a diff 1`] = `
"expected
<div>
  <div class=\\"label\\">Hello:</div>
  <div class=\\"value\\" data-test=\\"value\\">Jane Doe</div>
</div>
to satisfy <div><div class=\\"label\\">Hello:</div><div>Jane Doe!</div></div>

<div>
  <div class=\\"label\\">Hello:</div>
  <div class=\\"value\\" data-test=\\"value\\">
    Jane Doe // should equal 'Jane Doe!'
             //
             // -Jane Doe
             // +Jane Doe!
  </div>
</div>"
`;
